### PR TITLE
Tutorial 4: apiserver

### PR DESCRIPTION
**Pull request by stillinbeta**
/cc @lizthegrey

### CODE REVIEW

File: mdspell.yaml [hunk 1]
```diff
@@ -12,12 +13,15 @@ ignored:
 - kubelet
 - kubelet
 - kubernetes
```
Comment (lizthegrey) [inline mdspell.yaml:14]: "kubernetes" lowercase is probably a spelling error, but "Kubernetes" uppercase should be in the dictionary.

File: mdspell.yaml [hunk 2]
```diff
@@ -12,12 +13,15 @@ ignored:
 - kubelet
 - kubelet
```
Comment (lizthegrey) [inline mdspell.yaml]: appears twice.

File: tutorial4.md [hunk 3]
```diff
@@ -0,0 +1,424 @@
+# Tutorial 4: Getting the API server running
+
+The API server is the brain of kubernetes.
```
Comment (lizthegrey) [inline tutorial4.md]: Capitalize Kubernetes throughout.

File: tutorial4.md [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+It's the only component that communicates directly with `etcd`,
+and usually the only publicly exposed port.
```
Comment (lizthegrey) [inline tutorial4.md]: Sorry, this bit about the only publicly exposed port doesn't make sense -- on the entire cluster? on each VM?

File: mdspell.yaml [hunk 5]
```diff
@@ -4,6 +4,7 @@ ignored:
 - etcd
```
Comment (lizthegrey) [inline mdspell.yaml:4]: API but not api.

File: tutorial4.md [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+All kubernetes clients interact pretty much exclusively with the API server, including proxy ports with `kubectl port-forward` and running tasks with `kubectl exec`.
+
+By the end of this tutorial, your api server will be stood up and this server is going to look a lot more like a Kubernetes master node.
```
Comment (lizthegrey) [inline tutorial4.md]: API server

File: tutorial4.md [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+]
+$ openssl verify -CAfile /etc/kubernetes/pki/ca.crt  /etc//kubernetes/pki/apiserver.crt
+/etc//kubernetes/pki/apiserver.crt: OK
+```
+
+_NOTE: the IP addresses and names depend on your environment and hostname._
+
+### API Server etcd client
+
+API Server to to talk to etcd as well, so we'll generate credentials that are valid with etcd.
```
Comment (lizthegrey) [inline tutorial4.md]: needs to?

File: tutorial4.md [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+
+```console
+$ sudo kubeadm init phase certs apiserver-etcd-client
+[certs] Generating "apiserver-etcd-client" certificate and key
+$ openssl verify -CAfile /etc/kubernetes/pki/etcd/ca.crt /etc/kubernetes/pki/apiserver-etcd-client.crt
```
Comment (lizthegrey) [inline tutorial4.md:62]: we should ask people to note that etcd CA is used here, and challenge them on what happens if they verify against the non-etcd cert.
```console
elizabeth@lily:~$ openssl verify -CAfile /etc/kubernetes/pki/ca.crt /etc/kubernetes/pki/apiserver-etcd-client.crt
O = system:masters, CN = kube-apiserver-etcd-client
error 20 at 0 depth lookup: unable to get local issuer certificate
error /etc/kubernetes/pki/apiserver-etcd-client.crt: verification failed
```

File: tutorial4.md [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+    - --etcd-servers=https://127.0.0.1:2379
+    - --tls-cert-file=/etc/kubernetes/pki/apiserver.crt
+    - --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
+    - --secure-port=6443
+    - --insecure-port=0
+```
+
+#### Authentication
+
+We are going to be client certificate authentication.
```
Comment (lizthegrey) [inline tutorial4.md]: be doing

File: tutorial4.md [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+```
+
+If all's well, you should now be able to see the pods the kubelet is running for you:
+
+```shell
+$ kubectl get pods --namespace kube-system
+NAME                          READY   STATUS    RESTARTS   AGE
+etcd-k8s-tutorial             1/1     Running   0          2m10s
+kube-apiserver-k8s-tutorial   1/1     Running   0          2m10s
+```
```
Comment (lizthegrey) [inline tutorial4.md:432]: "if you're curious, you can look at the contents of `/etc/kubernetes/kubelet.conf` now that it's created (it's not necessary for `/etc/kubernetes/bootsrap-kubelet.conf` to exist as long as kubelet.conf exists)..."

File: tutorial4.md [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+    - --etcd-cafile=/etc/kubernetes/pki/etcd/ca.crt
+    - --etcd-certfile=/etc/kubernetes/pki/apiserver-etcd-client.crt
+    - --etcd-keyfile=/etc/kubernetes/pki/apiserver-etcd-client.key
+    - --etcd-servers=https://127.0.0.1:2379
+    - --tls-cert-file=/etc/kubernetes/pki/apiserver.crt
+    - --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
+    - --secure-port=6443
+    - --insecure-port=0
+    - --client-ca-file=/etc/kubernetes/pki/ca.crt
+    - --authorization-mode=Node,RBAC
```
Comment (stillinbeta) [inline tutorial4.md:292]: Need to add `--disable-admission-plugins=ServiceAccount ` or the whole thing falls apart later

File: mdspell.yaml (hunk 1 above)
Comment (stillinbeta) [reply mdspell.yaml:14]: https://github.com/mtuchowski/mdspell/blob/master/lib/mdspell/spell_checker.rb#L60 unfortunately this spell checker is case-insensitive :(

File: tutorial4.md (hunk 11 above)
Comment (stillinbeta) [reply tutorial4.md:292]: will add this as part of #7

**Pull request merged**